packages/bvaughn-architecture-demo/src/suspense/PauseCache.ts

```diff
@@ -1,214 +1,99 @@
-
-    wakeable.reject(error);
   }
-}
-
-async function fetchPauseData(
-  client: ReplayClientInterface,
-  pauseId: PauseId,
-  record: Record<PauseData>
-) {
-  const wakeable = record.value as Wakeable<PauseData>;
-
-  try {
-    const pauseData = await client.getAllFrames(pauseId);
-
-    // Pre-cache object previews that cam back with new Pause data.
-    // This will avoid us having to turn around and request them again when rendering the objects.
-    if (pauseData.objects) {
-      preCacheObjects(pauseId, pauseData.objects);
+  if (pauseData.scopes) {
+    for (const scope of pauseData.scopes) {
+      cacheScope(scope, pauseId, scope.scopeId);
     }
-
-    record.status = STATUS_RESOLVED;
-    record.value = pauseData;
-
-    wakeable.resolve(record.value);
-  } catch (error) {
-    record.status = STATUS_REJECTED;
-    record.value = error;
-
-    wakeable.reject(error);
   }
 }
 
-async function fetchPauseId(
+export function sortFramesAndUpdateLocations(
   client: ReplayClientInterface,
-  executionPoint: ExecutionPoint,
-  record: Record<Pause>
+  rawFrames: Frame[],
+  stack: FrameId[]
 ) {
-  const wakeable = record.value as Wakeable<Pause>;
-
-  try {
-    const pause = await client.createPause(executionPoint);
-
-    // Pre-cache object previews that cam back with new Pause data.
-    // This will avoid us having to turn around and request them again when rendering the objects.
-    if (pause.data.objects) {
-      preCacheObjects(pause.pauseId, pause.data.objects);
+  const frames = stack.map(frameId => rawFrames?.find(frame => frame.frameId === frameId));
+  if (frames.every(frame => !!frame)) {
+    for (const frame of frames) {
+      updateMappedLocation(client, frame!.location);
+      if (frame!.functionLocation) {
+        updateMappedLocation(client, frame!.functionLocation);
+      }
     }
-
-    record.status = STATUS_RESOLVED;
-    record.value = pause;
-
-    trackExecutionPointPauseIds(executionPoint, pause.pauseId);
-
-    wakeable.resolve(record.value);
-  } catch (error) {
-    record.status = STATUS_REJECTED;
-    record.value = error;
-
-    wakeable.reject(error);
+    return frames as Frame[];
   }
 }
 
-export function trackExecutionPointPauseIds(
-  executionPoint: ExecutionPoint,
-  newPauseId: PauseId
-): void {
-  const firstPauseId = executionPointToPauseIdMap.get(executionPoint);
-  if (firstPauseId == null) {
-    executionPointToPauseIdMap.set(executionPoint, newPauseId);
-  } else if (firstPauseId !== newPauseId) {
-    const error = new Error(
-      `Point (${executionPoint}) has multiple pause ids (${firstPauseId}, ${newPauseId})`
-    );
-
-    // TODO Hook this up so that it fails our e2e tests if called.
-    captureException(error);
-
-    console.error(error);
+function updateMappedLocation(client: ReplayClientInterface, mappedLocation: MappedLocation) {
+  for (const location of mappedLocation) {
+    location.sourceId = client.getCorrespondingSourceIds(location.sourceId)[0];
```

Comment:
I'm a little uneasy about _mutating_ data. That can lead to unpredictable/inconsistent results.

Can this function return a new location? Can we update `sortFramesAndUpdateLocations` return a new array of new location objects (rather than a new array of mutated location objects)?